src/main/scala/io/iohk/ethereum/blockchain/sync/SyncStateSchedulerActor.scala

```diff
@@ -43,135 +86,249 @@ class SyncStateSchedulerActor(downloader: ActorRef, sync: SyncStateScheduler, sy
+  ): ProcessingResult = {
+    requestResult match {
+      case RequestData(nodeData, from) =>
+        val (resp, newDownloaderState) = currentState.currentDownloaderState.handleRequestSuccess(from, nodeData)
+        resp match {
+          case UnrequestedResponse =>
+            ProcessingResult(
+              Left(DownloaderError(newDownloaderState, from, "unrequested response", blacklistPeer = false))
+            )
+          case NoUsefulDataInResponse =>
+            ProcessingResult(
+              Left(DownloaderError(newDownloaderState, from, "no useful data in response", blacklistPeer = true))
+            )
+          case UsefulData(responses) =>
+            sync.processResponses(currentState.currentSchedulerState, responses) match {
+              case Left(value) =>
+                ProcessingResult(Left(Critical(value)))
+              case Right((newState, stats)) =>
+                ProcessingResult(
+                  Right(ProcessingSuccess(newState, newDownloaderState, currentState.currentStats.addStats(stats)))
+                )
+            }
+        }
+      case RequestFailed(from, reason) =>
+        log.debug("Processing failed request from {}. Failure reason {}", from, reason)
+        val newDownloaderState = currentState.currentDownloaderState.handleRequestFailure(from)
+        ProcessingResult(Left(DownloaderError(newDownloaderState, from, reason, blacklistPeer = true)))
+    }
+  }
+
+  private def handleRestart(
       currentState: SchedulerState,
       currentStats: ProcessingStatistics,
       targetBlock: BigInt,
-      syncInitiator: ActorRef
-  ): Receive = {
-    case MissingNodes(nodes, downloaderCap) =>
-      log.debug(s"Received {} new nodes to process", nodes.size)
-      // Current SyncStateDownloaderActor makes sure that there is no not requested or duplicated values in its response.
-      // so we can ignore those errors.
-      //TODO [ETCM-275] process responses asynchronously to avoid steep rise of pending requests after pivot block update
-      sync.processResponses(currentState, nodes) match {
-        case Left(value) =>
-          log.error(s"Critical error while state syncing ${value}, stopping state sync")
-          // TODO we should probably start sync again from new target block, as current trie is malformed or declare
-          // fast sync as failure and start normal sync from scratch
-          context.stop(self)
-        case Right((newState, statistics)) =>
-          if (newState.numberOfPendingRequests == 0) {
-            reportStats(syncInitiator, currentStats, currentState)
-            finalizeSync(newState, targetBlock, syncInitiator)
-          } else {
+      restartRequester: ActorRef
+  ): Unit = {
+    log.debug("Starting request sequence")
+    sync.persistBatch(currentState, targetBlock)
+    restartRequester ! WaitingForNewTargetBlock
+    context.become(idle(currentStats.addSaved(currentState.memBatch.size)))
+  }
+  import akka.pattern.pipe
```

Comment:
very minor - I'd prefer to have this import either locally within method or on top of the file